internal/mutators/podtemplatespec.go

```diff
@@ -0,0 +1,118 @@
+package mutators
+
+import (
+	"fmt"
+
+	"github.com/kanopy-platform/hedgetrimmer/internal/limitrange"
+	"github.com/kanopy-platform/hedgetrimmer/internal/quantity"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+type PodTemplateSpec struct {
+	defaultMemoryLimitRequestRatio float64
+}
+
+func NewPodTemplateSpec(defaultMemoryLimitRequestRatio float64) PodTemplateSpec {
+	return PodTemplateSpec{
+		defaultMemoryLimitRequestRatio: defaultMemoryLimitRequestRatio,
+	}
+}
+
+func (p PodTemplateSpec) Mutate(inputPts corev1.PodTemplateSpec, limitRangeMemory limitrange.Config) (corev1.PodTemplateSpec, error) {
+	pts := *inputPts.DeepCopy()
+
+	if err := p.setAndValidateResourceRequirements(pts.Spec.InitContainers, limitRangeMemory); err != nil {
+		return pts, err
+	}
+
+	if err := p.setAndValidateResourceRequirements(pts.Spec.Containers, limitRangeMemory); err != nil {
+		return pts, err
+	}
+
+	return pts, nil
+}
+
+func (p PodTemplateSpec) setAndValidateResourceRequirements(containers []corev1.Container, limitRangeMemory limitrange.Config) error {
+	for idx := range containers {
+		container := &containers[idx]
+		p.setMemoryRequest(container, limitRangeMemory)
+
+		if err := p.setMemoryLimit(container, limitRangeMemory); err != nil {
+			return err
+		}
+
+		if err := p.validateMemoryRequirements(*container, limitRangeMemory); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p PodTemplateSpec) validateMemoryRequirements(container corev1.Container, limitRangeMemory limitrange.Config) error {
+	memoryRequest := container.Resources.Requests.Memory()
+	memoryLimit := container.Resources.Limits.Memory()
+
+	if memoryRequest.IsZero() || memoryLimit.IsZero() {
+		return fmt.Errorf("memory request (%s) and limit (%s) must be set", memoryRequest.String(), memoryLimit.String())
+	}
+
+	if memoryLimit.Cmp(*memoryRequest) == -1 {
+		return fmt.Errorf("memory limit (%s) must be greater than request (%s)", memoryLimit.String(), memoryRequest.String())
+	}
+
+	if limitRangeMemory.HasMaxLimitRequestRatio {
+		ratio := quantity.Div(*memoryLimit, *memoryRequest, infScaleMicro)
+		if ratio.Cmp(limitRangeMemory.MaxLimitRequestRatio) == 1 {
+			return fmt.Errorf("memory limit (%s) to request (%s) ratio (%s) exceeds MaxLimitRequestRatio (%s)",
+				memoryLimit.String(), memoryRequest.String(), ratio.String(), limitRangeMemory.MaxLimitRequestRatio.String())
+		}
+	}
+
+	return nil
+}
+
+func (p PodTemplateSpec) setMemoryRequest(container *corev1.Container, limitRangeMemory limitrange.Config) {
+	memoryRequest := container.Resources.Requests.Memory()
+	if !memoryRequest.IsZero() || !limitRangeMemory.HasDefaultRequest {
+		return
+	}
```

Comment:
Per the [k8s docs](https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/#what-if-you-specify-a-container-s-limit-but-not-its-request), if a container passes a limit but no request, the request is assigned to the same value as the limit. That behavior should be reflected here since it would be valid from limitranger standpoint. We would probably assign the value in this order.

1. user supplied limit value
2. limit range supplied default limit value
3. unset

Reply:
Good catch, had to think about this one a bit because consider these scenarios:
Input -> Output

```
request: unset, limit: 5Gi
request: 5Gi, limit: 5Gi
```

```
request: unset, limit: unset
(default request and limit exists 50Mi 64Mi, use them)
request: 50Mi, limit: 64Mi
```

```
request: unset, limit: unset
(only default limit 64Mi exists, use that for both, because the Limit gets set later, and request should match)
request: 64Mi, limit: 64Mi
```

So the evaluation order becomes:
- user supplied limit
- default request
- limit range supplied default limit
- unset

unit tests updated to reflect above